Convert unittests in `.d` files to use d-lang unittests

Alex-Muirhead opened the pull request:
## Motivation

Current unittests in D source files use `tcl` scripts for testing, and generate a separate executable for each unittest (enabled using specific version flags). This leaves 3 points of failure: 

1. Not creating a recipe for the specific unittest
2. Not enabling the correct version flags for the specific unittest
3. Not adding the target to the `tcl` script for testing

By converting all unittests to use the specific `unittest` block (enabled via the `--unittest` flag), we can compile an entire module to a single executable that runs all tests. As long as the file is marked as a source file, any unittests it contains will be correctly run.

## Additions

However, the default test-runner for D is extremely noisy (listing entire stack traces for assertion failures), without adding much information.

<details><summary>Example verbose output</summary>
<br>

For a single assert fail:
```text
core.exception.AssertError@equilibrium_update.d(370): Assertion failure
----------------
??:? [0x574b2e669c9e]
??:? [0x574b2e669902]
??:? [0x574b2e68eb8e]
??:? [0x574b2e67173c]
??:? [0x574b2e668952]
equilibrium_update.d:370 [0x574b2e5f9a1d]
??:? [0x574b2e669bb7]
??:? [0x574b2e6748d6]
??:? [0x574b2e674e08]
??:? [0x574b2e67487b]
??:? [0x574b2e66e1ee]
??:? [0x574b2e669a63]
??:? [0x574b2e67142a]
??:? [0x574b2e671357]
??:? [0x574b2e6711ac]
/home/alex/.dlang/ldc2-v1.41.0/bin/../import/core/internal/entrypoint.d:42 [0x574b2e4337a1]
??:? [0x71a47902a1c9]
??:? __libc_start_main [0x71a47902a28a]
??:? [0x574b2e4336a4]
33/66 modules FAILED unittests
```

</details>

I've added a separate `test_runner.d` file to `src/utils/` that aims to emulate the output of PyTest (already used in `examples/`).

<details open><summary>Example test_runner output</summary>
<br>

```text
============================= test session starts =============================
collected 9 test modules

kinetics.chemistry_update::unittest PASSED
kinetics.equilibrium_update::unittest FAILED
kinetics.rate_constant::unittest PASSED
kinetics.reaction::unittest PASSED
kinetics.reaction_mechanism::unittest PASSED
kinetics.two_temperature_argon_kinetics::unittest PASSED
kinetics.two_temperature_argon_with_ideal_gas::unittest SKIPPED
kinetics.two_temperature_gasgiant_kinetics::unittest PASSED
kinetics.vib_specific_co_kinetics::unittest PASSED

================================== FAILURES ===================================
____________________ kinetics.equilibrium_update::unittest ____________________
T: 2499.99
  equilibrium_update.d:L370 - Assertion failure

=================== 1 failed, 7 passed, 1 skipped in 0.07s ====================
```

</details>

## Features

- **Module filtering:** By default, _all_ files added to the compiler call (with `--unittest`) will have their unittest run. This is usually more information than required. The executable generated by `test_runner.d` has a `--module <name>` option, which filters tests down to those belonging to modules starting with `<name>` (i.e. `--module gas.diffusion` will only run tests within `gas/diffusion/`). 
- **`stdout` capturing:** The previous unittesting would fail a test if it wrote to `stdout` (via `writeln` etc). Within the new framework, any output is captured, and only displayed in the error description of failed/errored tests. This helps with debugging tests without cluttering the output / causing false positives.
- **Skipped tests:** In some cases, a test is known to not succeed, but the failure is well known and not causing any immediate problems. The `util.test_runner` module contains a `skip(string message = "Skipped test")` function, and the relevant unittest won't count as failed or errored. This function won't cause any problems with existing test runners.

## Notes

- As an entire module is compiled to a single executable, two separate compilations are needed if complex numbers are tested separately. Currently, the `makefile`s for these modules contain combinations of `test`, `test-all`, `test-real`, and `test-complex` targets. A unified naming convention would be best.

Alex-Muirhead committed:
gas/diffusion: Fix input args to GasMixture

Alex-Muirhead committed:
geom: Convert geom to use d-lang unittest

Alex-Muirhead committed:
kinetics: Convert kinetics to use d-lang unittest

Alex-Muirhead committed:
gas: Create integration test for gas-calc
To make this play nicer with our existing d-unittests, we create a separate test_gas_calc file, and ensure we build gas-calc before running the tests.
This new test_gas_calc uses std.process to launch the compiled gas-calc.

Alex-Muirhead committed:
gas: Convert gas to use d-lang unittest

Alex-Muirhead committed:
ntypes: Convert ntypes to d-lang unittests

Alex-Muirhead committed:
ntypes: Move ntypes unittest within version block

Alex-Muirhead committed:
gas: Add missing chdir to test

rjgollan-on-github commented on src/nm/makefile:
```diff
@@ -79,151 +60,36 @@ DFLAGS += -dip1008 -preview=in
+test_runner: $(NM_FILES) $(NM_LUA_FILES) $(NTYPES_FILES) $(UTIL_FILES) $(LIBLUA)
+	@echo "Compiling test runner..."
+	$(DMD) -of$@ -debug -g -w -dip1008 --unittest $(DLINKFLAGS) \
+		$(filter %/test_runner.d,$^) $(filter-out %/test_runner.d,$^)
```
Explicitly list test_runner first to match intention.

Alex-Muirhead committed:
src: Explicitly put test_runner.d as first file
This makes the intention clearer than using a filter operation,
however runs the risk of the relative path being incorrect if moved.

Alex-Muirhead committed:
geom/volume: Fix formatting on multi-d array

Alex-Muirhead committed:
src: Add explicit dependencies to test_runner.d

rjgollan-on-github commented:
I looked over this, but didn't action.

I'm happy with the ".d" style unittests.

Nick, can you please sense check then merge if you're happy?

On Fri, 13 Feb 2026, 4:07 pm Nick Gibbons, ***@***.***> wrote:

> *uqngibbo* left a comment (gdtk-uq/gdtk#141)
> <https://github.com/gdtk-uq/gdtk/pull/141#issuecomment-3894559747>
>
> Is this waiting for me Alex? Sorry I thought we had merged this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gdtk-uq/gdtk/pull/141#issuecomment-3894559747>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWONKBYOTME3LM7WWZCX3VL4LU5WZAVCNFSM6AAAAACG52I3RSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTQOJUGU2TSNZUG4>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>
